File: tests/src/test/java/client/CryptographyTests.java
package client;

import com.github.blahblahbloopster.crypto.Crypto;
import net.i2p.crypto.eddsa.EdDSAPrivateKey;
import net.i2p.crypto.eddsa.EdDSAPublicKey;
import org.junit.jupiter.api.*;
import java.security.KeyPair;
import java.util.Random;
import static org.junit.jupiter.api.Assertions.*;

/** A few tests for {@link Crypto} */
public class CryptographyTests {

    @BeforeAll
    static void init(){
        Crypto.INSTANCE.init();
    }

    /** Tests signature creation and validation. */
    @Test
    void testSigning() {
        KeyPair pair = Crypto.INSTANCE.generateKeyPair();

        byte[] input = new byte[128];
        new Random().nextBytes(input);

        byte[] signature = Crypto.INSTANCE.sign(input, pair.getPrivate());

        assertTrue(Crypto.INSTANCE.verify(input, signature, (EdDSAPublicKey) pair.getPublic()));
    }

    /** Tests key serialization and deserialization. */
    @Test
    void testSerialization() {
        KeyPair pair = Crypto.INSTANCE.generateKeyPair();

        byte[] encodedPublic = Crypto.INSTANCE.serializePublic((EdDSAPublicKey) pair.getPublic());
        byte[] encodedPrivate = Crypto.INSTANCE.serializePrivate((EdDSAPrivateKey) pair.getPrivate());

        assertEquals(Crypto.INSTANCE.deserializePublic(encodedPublic), pair.getPublic());
        assertEquals(Crypto.INSTANCE.deserializePrivate(encodedPrivate), pair.getPrivate());
    }
}

File: tests/src/test/java/client/CryptographyTests.kt
package client

import mindustry.client.crypto.Crypto
import mindustry.client.crypto.CryptoClient
import mindustry.client.crypto.KeyQuad
import mindustry.client.crypto.PublicKeyPair
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import kotlin.random.Random

/** A few tests for [Crypto]  */
class CryptographyTests {

    companion object {
        @BeforeAll
        @JvmStatic
        fun init() {
            Crypto.initializeAlways()
        }
    }

    /** Tests key serialization and deserialization.  */
    @Test
    fun testSerialization() {
        val quad = Crypto.generateKeyQuad()
        val decoded = KeyQuad(quad.serialize())
        assertEquals(quad, decoded)

        val pair = PublicKeyPair(quad)
        val decodedPair = PublicKeyPair(pair.serialize())
        assertEquals(pair, decodedPair)
    }

    @Test
    fun testClientInteractions() {
        val client1 = CryptoClient(Crypto.generateKeyQuad())
        val client2 = CryptoClient(Crypto.generateKeyQuad())

        client1.generate(PublicKeyPair(client2.key.xPublicKey, client2.key.edPublicKey))
        client2.generate(PublicKeyPair(client1.key.xPublicKey, client1.key.edPublicKey))

        assertTrue(client1.sharedSecret.contentEquals(client2.sharedSecret))
        assertEquals(client1.aesKey, client2.aesKey)
        assertTrue(client1.iv.iv.contentEquals(client2.iv.iv))

        val original = Random.nextBytes(1024)
        assertTrue(client1.verify(client2.sign(original), original))
        assertTrue(client2.verify(client1.sign(original), original))

        val plaintext = Random.nextBytes(1024)
        val encrypted1 = client1.encrypt(plaintext)
        val encrypted2 = client2.encrypt(plaintext)

        assertTrue(client2.decrypt(encrypted1).contentEquals(plaintext))
        assertTrue(client1.decrypt(encrypted2).contentEquals(plaintext))
    }
}
